### PR TITLE
Paper mods2: added load-imbalance timers

### PR DESCRIPTION
**Pull request by johnmauff**
This PR removes some dead code that was commented out and adds a new timer that measures load imbalance in the model.  It also eliminates a device-to-host copy that was unnecessary.  Note that to accurately measure load imbalance, a number of communication operators were grouped together  This also reduces the number of cpp  #ifdef _MPI in the code.

### CODE REVIEW

File: run/gust.node8.sh [hunk 1]
```diff
@@ -7,17 +7,12 @@
...
-module purge
-module load nvhpc
-module load cray-mpich
-module load ncarcompilers
-module unload cuda
+module --force purge
+module load ncarenv/23.03
+module load nvhpc/23.1
+module load ncarcompilers/0.8.0
+module load cray-mpich/8.1.25
```
Comment (sjsprecious) [inline run/gust.node8.sh:14]: For CPU run, shall we use the intel compiler instead for better performance?

File: src/cm1.F [hunk 2]
```diff
@@ -4381,10 +4381,19 @@ program cm1
         sum = 0.0
         call MPI_REDUCE(time_parceli ,sum,1,MPI_REAL,MPI_SUM,0,MPI_COMM_WORLD,ierr)
         time_parceli = sum/float(numprocs)
+
         sum = 0.0
         call MPI_REDUCE(time_droplet ,sum,1,MPI_REAL,MPI_SUM,0,MPI_COMM_WORLD,ierr)
         time_droplet = sum/float(numprocs)
```
Comment (sjsprecious) [inline src/cm1.F:4388]: Does `time_droplet` include the timing of `time_lb` and `time_droplet_inject` below?

File: src/droplet.F [hunk 3]
```diff
@@ -2167,6 +2177,7 @@ subroutine gauss_newton_2d(diffnorm,dt_nondim,taup_scale,rt_start,rt_zeros,flag,
                 v1(2) = rt_zeros(2)
 
         end do
+        !print *,'gauss_newton_2d: iterations: ',iterations
```
Comment (sjsprecious) [inline src/droplet.F]: remove this line

File: src/droplet.F [hunk 4]
```diff
@@ -2482,7 +2493,7 @@ subroutine droplet_injection(pdata,xf,yf,dt,s10_xym,randomNumbers)
 
       my_reintro = floor(xrange*yrange*dt*totdrops/drop_mult)      !Number of droplets to be produced by this MPI task
 
-      my_reintro = 10
+      !JMD my_reintro = 10
```
Comment (sjsprecious) [inline src/droplet.F]: remove this line (also removed in David's PR)

File: src/solve3.F [hunk 5]
```diff
@@ -405,13 +406,12 @@ subroutine solve3(nstep,rbufsz,num_soil_layers,                  &
       if(timestats.ge.1) time_prsrho=time_prsrho+mytime()
 
 
-        call bcs_GPU(rr)
+      call bcs_GPU(rr)
+      call bcs_GPU(rf)
```
Comment (sjsprecious) [inline src/solve3.F:410]: I saw that you moved many subroutine calls like this and I did not think they were answer changing. I am just curious about why you want to re-order these calling sequences and how you determine to move their locations up or down?

File: src/turb.F [hunk 6]
```diff
@@ -2498,42 +2487,35 @@ subroutine tkekm(nstep,rtime,dt,ruh,rvh,rmh,zf,mf,rmf,zntmp,ust,ustt,rf,cmz,csz,
...
+!    if( nstep.eq.0 .and. myid.eq.0 )then
+!      !$acc update host (grdscl)
+!      print *
+!      print *,'  zf,grdscl:'
+!      i = 1
+!      j = 1
+!      do k=2,nk
+!        print *,k,(zf(i,j,k)-zf(i,j,1)),grdscl(i,j,k)
+!      enddo
+!    endif
```
Comment (sjsprecious) [inline src/turb.F]: remove these lines

**Review by sjsprecious — COMMENTED** [review]
Thanks @johnmauff for issuing this PR. I did not run a verification test but I did not think these changes were answer changing with my visual judgement. Just have a few clarification questions.

File: run/gust.node8.sh (hunk 1 above)
Comment (johnmauff) [reply run/gust.node8.sh:14]: Jian, we could you if want.  While on Cheyenne the Intel compiler offers significant advantages versus nvhpc, however for the AMD processors, I don't see much difference between Intel and the nvhpc compiler.

File: src/cm1.F (hunk 2 above)
Comment (johnmauff) [reply src/cm1.F:4388]: No, the time_droplet is independent of either time_lb (Time spent in load imbalance) or the injection time.  I wanted to see how expensive the droplet injection calculation really was.  We can certainly remove the separate time_drolet_inject once we are done optimizing the entire Droplet calculation.

File: src/droplet.F (hunk 4 above)
Comment (johnmauff) [reply src/droplet.F]: Done

File: src/droplet.F (hunk 3 above)
Comment (johnmauff) [reply src/droplet.F]: Done

File: src/turb.F (hunk 6 above)
Comment (johnmauff) [reply src/turb.F]: Done

File: src/solve3.F (hunk 5 above)
Comment (johnmauff) [reply src/solve3.F:410]: The moving of subroutine calls around was associated with adding the load-imbalance timers, and the desire to reduce the number of #ifdef _MPI directives in the code.  So I grouped the packing routines bc*_GPU and the MPI boundary exchange calls together.

File: run/gust.node8.sh (hunk 1 above)
Comment (sjsprecious) [reply run/gust.node8.sh:14]: Oh really, I did not realize that. I am still under the impression that the intel compiler outperforms the nvhpc compiler significantly on the CPU but it looks like this is limited to the Intel's CPU only? This is important information to be recorded and tested in other models, too.

File: src/cm1.F (hunk 2 above)
Comment (sjsprecious) [reply src/cm1.F:4388]: Thanks for the clarification. Clear to me now.

File: src/solve3.F (hunk 5 above)
Comment (sjsprecious) [reply src/solve3.F:410]: So does it mean that it does not matter if I call the `comm_*_GPU` subroutines before or after the `bc*_GPU` subroutines, as long as I group them together first?

File: run/gust.node8.sh (hunk 1 above)
Comment (johnmauff) [reply run/gust.node8.sh:14]: I was actually surprised as well.  It is illustrated in Figure 5 of the CM1 paper.  I have not checked for other codes yet.

File: src/solve3.F (hunk 5 above)
Comment (johnmauff) [reply src/solve3.F:410]: Jian,  I kept the overall order between bc*_GPU and comm_*_GPU calls.  The order is critical, it just does not matter if all the operations on variable 'foo' are performed before the operations on variable 'bar'.  For example  

Old order 

call bcs_GPU(foo)
call comm_start_GPU(foo)
call bcs_GPU(bar)
call comm_start_GPU(bar)

New order:

call bcs_GPU(foo)
call bcs_GPU(bar)
call comm_start_GPU(foo)
call comm_start_GPU(bar)

Comment (sjsprecious) [reply src/solve3.F:410]: Thanks @johnmauff . That is clear to me now.

**Comment by sjsprecious** [comment]
By using the namelist `namelist.restC`, the verification result on Casper looks below:

---

CPU (intel, 2 MPI ranks) vs. GPU (nvhpc, 2 MPI ranks + 2 V100 GPUs)

Stat output comparision

Identical variables: 63
variables with rel_diff > 1.e-6: 23
variables with rel_diff <= 1.e-6: 0
four variables with largest error: PPMIN, PPMAX, VOR2KM, DIVMAX

Droplet 0th-order diagnostic comparison

Identical variables: 10
variable with rel_diff > 1.e-6: 2
variable with rel_diff <= 1.e-6: 0
Two variables with largest error: Tpavg (~3.4e-6), radmin (~6.2e-6)

Droplet 1st-order diagnostic comparison

Identical variables: 512
variable with rel_diff > 1.e-6: 0
variable with rel_diff <= 1.e-6: 0

---

The output is identical with that in PR #103.

**Review by sjsprecious — APPROVED** [review]

**Pull request merged**